File: java/retrieve heap/src/testsuites/Treep_Insert_Rec.java
package testsuites;

import java.util.Random;

public class Treep_Insert_Rec {
    static PNode create_tree(int size){
        Random rand = new Random();
        PNode root = null;
        for(int i = 0; i < size - 3; i++){
            root = treap_insert_node(root, i, rand.nextInt(20));
        }
        root = treap_insert_node(root, -5, rand.nextInt(20));
        root = treap_insert_node(root, -2, rand.nextInt(20));
        root = treap_insert_node(root, -4, rand.nextInt(20));
        return root;
    }

    static PNode treap_insert_node(PNode x, int k, int p)
    {
        if (x == null) {
            PNode leaf = new PNode();
            leaf.key = k;
            leaf.prio = p;
            leaf.left = null;
            leaf.right = null;
            return leaf;
        } else {
            if (k < x.key) {
                PNode xr = x.right;
                PNode xl = x.left;
                PNode tmp = treap_insert_node(xl, k, p);
                if (tmp.prio <= x.prio) {
                    x.left = tmp;
                    return x;
                } else {
                    PNode lr = tmp.right;
                    x.left = lr;
                    tmp.right = x;
                    return tmp;
                }
            } else {
                PNode xr = x.right;
                PNode xl = x.left;
                PNode tmp = treap_insert_node(xr, k, p);
                if (tmp.prio <= x.prio) {
                    x.right = tmp;
                    return x;
                } else {
                    PNode rl = tmp.left;
                    x.right = rl;
                    tmp.left = x;
                    return tmp;
                }
            }
        }
    }

    static PNode treap_insert_rec(PNode x, int k, int p)
    {
        //pre
        if (x == null) {
            PNode leaf = new PNode();
            leaf.key = k;
            leaf.prio = p;
            leaf.left = null;
            leaf.right = null;
            //post
            return leaf;
        } else {
            if (k < x.key) {
                PNode xr = x.right;
                PNode xl = x.left;
                PNode tmp = treap_insert_rec(xl, k, p);
                if (tmp.prio <= x.prio) {
                    x.left = tmp;
                    //post
                    return x;
                } else {
                    PNode lr = tmp.right;
                    x.left = lr;
                    tmp.right = x;
                    //post
                    return tmp;
                }
            } else {
                PNode xr = x.right;
                PNode xl = x.left;
                PNode tmp = treap_insert_rec(xr, k, p);
                if (tmp.prio <= x.prio) {
                    x.right = tmp;
                    //post
                    return x;
                } else {
                    PNode rl = tmp.left;
                    x.right = rl;
                    tmp.left = x;
                    //post
                    return tmp;
                }
            }
        }
    }

    public static void main(String args[]){
        Random rand = new Random();
        PNode root = create_tree(10);
        PNode res = treap_insert_rec(root, rand.nextInt(20),rand.nextInt(10));
        treap_insert_rec(root, -100,rand.nextInt(10));
        treap_insert_rec(root, 100,rand.nextInt(20));
    }
}

class PNode{
    int key;
    int prio;
    PNode left;
    PNode right;
}

File: benchmarks/python/treep_insert_rec.py
import random
import pprint

class node:
	def __init__(self, key = 0, prio = 0, left = None, right = None):
		self.key = key
		self.prio = prio
		self.left = left
		self.right = right

def insert_rec(x, k, p):
	if x is None:
		leaf = node(k, p)
		return leaf;
	elif k < x.key:
		xl = x.left
		xr = x.right
		tmp = insert_rec(xl, k, p)
		if tmp.prio <= x.prio:
			x.left = tmp
			return x
		else:
			lr = tmp.right
			x.left = lr
			tmp.right = xl
			return tmp
	else:
		xl = x.left
		xr = x.right
		tmp = insert_rec(xr, k, p)
		if tmp.prio <= x.prio:
			x.right = tmp
			return x
		else:
			rl = tmp.left
			x.right = rl
			tmp.left = xl
			return tmp

def create_tree(size):
	root = None
	for i in range(0,size-3):
		root = insert_rec(root, random.randint(0, 10), i)
	root = insert_rec(root, random.randint(0, 10), -5)
	root = insert_rec(root, random.randint(0, 10), -2)
	root = insert_rec(root, random.randint(0, 10), -4)
	return root

def insert_node(x, k, p):
	if x is None:
		leaf = node(k, p)
		return leaf;
	elif k < x.key:
		xl = x.left
		xr = x.right
		tmp = insert_rec(xl, k, p)
		if tmp.prio <= x.prio:
			x.left = tmp
			return x
		else:
			lr = tmp.right
			x.left = lr
			tmp.right = xl
			return tmp
	else:
		xl = x.left
		xr = x.right
		tmp = insert_rec(xr, k, p)
		if tmp.prio <= x.prio:
			x.right = tmp
			return x
		else:
			rl = tmp.left
			x.right = rl
			tmp.left = xl
			return tmp

def main():
	root = create_tree(5)
	insert_node(root, random.randint(0,10))
	insert_node(root, -100)
	insert_node(root, 100)
	insert_node(None, random.randint(0,10))

if __name__=="__main__":
	main()
